backend/checkmach.ml

```diff
@@ -1250,38 +1250,40 @@ end = struct
     | (Var _ | Join _ | Transform _), _ | _, (Var _ | Join _ | Transform _) ->
       assert false
 
-  let rec apply t ~env =
-    match t with
-    | Bot | Safe | Top _ -> t
-    | Var { var; witnesses } -> replace_witnesses witnesses (env var)
-    | Transform tr ->
-      let init =
-        match Transform.get_top tr with None -> Safe | Some w -> Top w
-      in
-      Vars.fold
-        ~f:(fun var w acc ->
-          let v = replace_witnesses w (env var) in
-          transform v acc)
-        ~init (Transform.get_vars tr)
-    | Join j ->
-      let init =
-        match Join.get_top j with
-        | None -> if Join.has_safe j then Safe else Bot
-        | Some w -> Top w
-      in
-      let vars, trs = Join.get j in
-      let acc =
+  let apply t ~env =
+    let env var w =
+      match env var with
+      | None -> unresolved w var
+      | Some v -> replace_witnesses w v
+    in
+    let rec aux t =
+      match t with
+      | Bot | Safe | Top _ -> t
+      | Var { var; witnesses } -> env var witnesses
```

Comment:
(If unresolved, will reallocate while we could return `t`.)

Reply:
Slightly improved `aux` to not reallocate if there is no change. Left a CR for further optimizations mentioned in your other comment.